astropy/units/quantity_helper/function_helpers.py

```diff
@@ -497,6 +501,63 @@ def arange_impl(*args, start=None, stop=None, step=None, dtype=None, device=None
     return np.arange(**kwargs), out_unit, None
 
 
+if NUMPY_LT_2_0:
+    @dispatched_function
+    def empty(shape, dtype=None, order='C'):
+        return empty_impl(shape, dtype, order)
+else:
+    @dispatched_function
+    def empty(shape, dtype=None, order='C', *, device=None):
+        return empty_impl(shape, dtype, order, device=device)
+
+def empty_impl(shape, dtype=None, order='C', *, device=None):
+    if NUMPY_LT_2_0:
+        ret_arr = np.empty(shape, dtype, order)
+    else:
+        ret_arr = np.empty(shape, dtype, order, device=device)
+    return ret_arr, _UnitFromLikeArg, None
+
+
+if NUMPY_LT_2_0:
+    @dispatched_function
+    def ones(shape, dtype=None, order='C'):
+        return ones_impl(shape, dtype, order)
+else:
+    @dispatched_function
+    def ones(shape, dtype=None, order='C', *, device=None):
+        return ones_impl(shape, dtype, order, device=device)
+
+def ones_impl(shape, dtype=None, order='C', *, device=None):
+    if NUMPY_LT_2_0:
+        ret_arr = np.ones(shape, dtype, order)
+    else:
+        ret_arr = np.ones(shape, dtype, order, device=device)
+    return ret_arr, _UnitFromLikeArg, None
+
+@dispatched_function
+def zeros(shape, dtype=None, order='C'):
+    return np.zeros(shape, dtype, order), _UnitFromLikeArg, None
+
+if NUMPY_LT_2_0:
+    @dispatched_function
+    def full(shape, fill_value, dtype=None, order='C'):
+        return full_impl(shape, fill_value, dtype, order)
+else:
+    @dispatched_function
+    def full(shape, fill_value, dtype=None, order='C', *, device=None):
+        return full_impl(shape, fill_value, dtype, order, device=device)
+
+def full_impl(shape, fill_value, dtype=None, order='C', *, device=None):
```

Comment:
Here one could still call the overall implementation, and then get the unit after.

Reply:
I don't feel comfortable generalizing early as I worry we could get trapped when later expanding the list of implemented function, so I'd like to avoid designing a general implementation for now if that's okay.